File: lib/audio_splitter.py
"""Audio splitting utilities"""

import subprocess
from pathlib import Path
from typing import List


def get_audio_duration(input_file: str) -> float:
    """
    Get duration of audio file in seconds

    Args:
        input_file: Full path to audio file

    Returns:
        Duration in seconds

    Raises:
        FileNotFoundError: If input file doesn't exist
        RuntimeError: If ffprobe fails
    """
    input_path = Path(input_file)
    if not input_path.exists():
        raise FileNotFoundError(f"Input file not found: {input_file}")

    cmd = [
        'ffprobe',
        '-v', 'error',
        '-show_entries', 'format=duration',
        '-of', 'default=noprint_wrappers=1:nokey=1',
        str(input_path)
    ]

    result = subprocess.run(
        cmd,
        capture_output=True,
        text=True
    )

    if result.returncode != 0:
        raise RuntimeError(f"ffprobe failed: {result.stderr}")

    try:
        duration = float(result.stdout.strip())
        return duration
    except ValueError:
        raise RuntimeError(f"Invalid duration output: {result.stdout}")


def split_audio(
    input_file: str,
    num_parts: int,
    output_folder: str,
    output_prefix: str = ""
) -> List[str]:
    """
    Split audio file into multiple parts

    Args:
        input_file: Full path to input audio file
        num_parts: Number of parts to split into
        output_folder: Folder where split files will be saved
        output_prefix: Prefix for output files (defaults to input filename)

    Returns:
        List of full paths to split audio files

    Raises:
        FileNotFoundError: If input file doesn't exist
        ValueError: If num_parts < 1
        RuntimeError: If ffmpeg split fails
    """
    input_path = Path(input_file)
    if not input_path.exists():
        raise FileNotFoundError(f"Input file not found: {input_file}")

    if num_parts < 1:
        raise ValueError("num_parts must be at least 1")

    # Create output folder
    output_path = Path(output_folder)
    output_path.mkdir(parents=True, exist_ok=True)

    # Use input filename if no prefix specified
    if not output_prefix:
        output_prefix = input_path.stem

    # Get duration
    duration = get_audio_duration(input_file)
    part_duration = int(duration / num_parts)

    split_files = []

    # Split the file
    for i in range(1, num_parts + 1):
        start_time = (i - 1) * part_duration
        output_file = output_path / f"{output_prefix}_part{i:02d}{input_path.suffix}"

        if i == num_parts:
            # Last part: go to end of file
            cmd = [
                'ffmpeg',
                '-i', str(input_path),
                '-ss', str(start_time),
                '-c', 'copy',
                '-y',
                str(output_file)
            ]
        else:
            # Other parts: specify duration
            cmd = [
                'ffmpeg',
                '-i', str(input_path),
                '-ss', str(start_time),
                '-t', str(part_duration),
                '-c', 'copy',
                '-y',
                str(output_file)
            ]

        result = subprocess.run(
            cmd,
            capture_output=True,
            text=True
        )

        if result.returncode != 0:
            raise RuntimeError(f"FFmpeg split failed for part {i}: {result.stderr}")

        if not output_file.exists():
            raise RuntimeError(f"Split failed: part {i} not created")

        split_files.append(str(output_file))

    return split_files

File: src/pipeline/audio-splitter.ts
import { execa } from "execa";
import { existsSync } from "node:fs";
import { mkdir } from "node:fs/promises";
import { basename, extname, join } from "node:path";

/**
 * Get audio duration in seconds using ffprobe.
 * Port of lib/audio_splitter.py get_audio_duration()
 */
export async function getAudioDuration(inputFile: string): Promise<number> {
  if (!existsSync(inputFile)) {
    throw new Error(`Input file not found: ${inputFile}`);
  }

  const result = await execa("ffprobe", [
    "-v", "error",
    "-show_entries", "format=duration",
    "-of", "default=noprint_wrappers=1:nokey=1",
    inputFile,
  ]);

  const duration = parseFloat(result.stdout.trim());
  if (isNaN(duration)) {
    throw new Error(`Invalid duration output: ${result.stdout}`);
  }

  return duration;
}

/**
 * Split audio file into multiple equal-length parts.
 * Port of lib/audio_splitter.py split_audio()
 */
export async function splitAudio(
  inputFile: string,
  numParts: number,
  outputFolder: string,
  outputPrefix?: string,
): Promise<string[]> {
  if (!existsSync(inputFile)) {
    throw new Error(`Input file not found: ${inputFile}`);
  }
  if (numParts < 1) {
    throw new Error("numParts must be at least 1");
  }

  await mkdir(outputFolder, { recursive: true });

  const ext = extname(inputFile);
  const stem = basename(inputFile, ext);
  const prefix = outputPrefix || stem;

  const duration = await getAudioDuration(inputFile);
  const partDuration = Math.floor(duration / numParts);

  const splitFiles: string[] = [];

  for (let i = 1; i <= numParts; i++) {
    const startTime = (i - 1) * partDuration;
    const partNum = String(i).padStart(2, "0");
    const outputFile = join(outputFolder, `${prefix}_part${partNum}${ext}`);

    const args = [
      "-i", inputFile,
      "-ss", String(startTime),
    ];

    // Last part: go to end of file; others: specify duration
    if (i < numParts) {
      args.push("-t", String(partDuration));
    }

    args.push("-c", "copy", "-y", outputFile);

    await execa("ffmpeg", args);

    if (!existsSync(outputFile)) {
      throw new Error(`Split failed: output file not created for part ${i}`);
    }

    splitFiles.push(outputFile);
  }

  return splitFiles;
}
